Introduce option to opt-out of spinners

BagToad opened the pull request:
### Description

Introduce `GH_SPINNER_DISABLED` environment variable to opt-out of displaying a "spinner" animation as a progress indicator. Instead, print `Working...` and any other contextual progress indication (AKA the "label").

### Examples

```
# No added label / prefix
gh search prs  test  -R  cli/cli  -L 300 
Working...

# Additional label / prefix
gh release download -R cli/cli --archive zip
Finding assets to download...
Downloading ...
```

![out](https://github.com/user-attachments/assets/e7035bd6-5731-4506-8fb4-a28038e7c54a)

BagToad committed:
feat(iostreams): Option to opt out of spinners

iamazeem commented:
Instead of `Working...`, the respective operation e.g.:

- `Searching...`
- `Finding assets to download...`
- `Downloading...`
- etc.

seem more appropriate here.

andyfeller commented on pkg/iostreams/iostreams.go:
```diff
@@ -294,9 +295,20 @@ func (s *IOStreams) StartProgressIndicatorWithLabel(label string) {
+		spinnerStyle = spinner.CharSets[11]
+	}
+
+	sp := spinner.New(spinnerStyle, 120*time.Millisecond, spinner.WithWriter(s.ErrOut), spinner.WithColor("fgCyan"))
```
**thought:** Even with these changes, I think `briandowns/spinner` is still erasing content and reprinting it as the underlying `erase()` logic doesn't make assumptions of what is on the screen and erases part of it.

Source: https://github.com/briandowns/spinner/blob/55430861f77b20de3f456775b00c706519464f36/spinner.go#L451C1-L481C1

```go
// erase deletes written characters on the current line.
// Caller must already hold s.lock.
func (s *Spinner) erase() {
	n := utf8.RuneCountInString(s.lastOutputPlain)
	if runtime.GOOS == "windows" && !isWindowsTerminalOnWindows {
		clearString := "\r" + strings.Repeat(" ", n) + "\r"
		fmt.Fprint(s.Writer, clearString)
		s.lastOutputPlain = ""
		return
	}

	numberOfLinesToErase := computeNumberOfLinesNeededToPrintString(s.lastOutputPlain)

	// Taken from https://en.wikipedia.org/wiki/ANSI_escape_code:
	// \r     - Carriage return - Moves the cursor to column zero
	// \033[K - Erases part of the line. If n is 0 (or missing), clear from
	// cursor to the end of the line. If n is 1, clear from cursor to beginning
	// of the line. If n is 2, clear entire line. Cursor position does not
	// change.
	// \033[F - Go to the beginning of previous line
	eraseCodeString := strings.Builder{}
	// current position is at the end of the last printed line. Start by erasing current line
	eraseCodeString.WriteString("\r\033[K") // start by erasing current line
	for i := 1; i < numberOfLinesToErase; i++ {
		// For each additional lines, go up one line and erase it.
		eraseCodeString.WriteString("\033[F\033[K")
	}
	fmt.Fprint(s.Writer, eraseCodeString.String())
	s.lastOutputPlain = ""
}
```

BagToad replied:
Resolving this comment to keep the conversation about this redrawing in one place under https://github.com/cli/cli/pull/10773#discussion_r2042622266

andyfeller commented on pkg/iostreams/iostreams.go:
```diff
@@ -294,9 +295,20 @@ func (s *IOStreams) StartProgressIndicatorWithLabel(label string) {
-	// https://github.com/briandowns/spinner#available-character-sets
-	dotStyle := spinner.CharSets[11]
-	sp := spinner.New(dotStyle, 120*time.Millisecond, spinner.WithWriter(s.ErrOut), spinner.WithColor("fgCyan"))
+	spinnerDisabledValue, spinnerDisabledIsSet := os.LookupEnv("GH_SPINNER_DISABLED")
```
**suggest:** Though there are some places where the environment variables are evaluated JIT, I would consider whether this should be moved to the IOStream initialization logic in the factory:

https://github.com/cli/cli/blob/408e21ebdddf9cd14289e49135389a6e5125eff4/pkg/cmd/factory/default.go#L273-L310

BagToad replied:
I did an attempt at this in 3eba461afb2746576cd26ed391683494c4f48dd5 :)

andyfeller commented on pkg/iostreams/iostreams.go:
```diff
@@ -294,9 +295,20 @@ func (s *IOStreams) StartProgressIndicatorWithLabel(label string) {
+	spinnerDisabledValue, spinnerDisabledIsSet := os.LookupEnv("GH_SPINNER_DISABLED")
+	falseyValues := []string{"false", "0", "no", ""}
+
+	var spinnerStyle []string
+	if spinnerDisabledIsSet && !slices.Contains(falseyValues, spinnerDisabledValue) {
+		// This progress indicator must be readable by screen readers
+		spinnerStyle = []string{"Working..."}
+	} else {
+		// https://github.com/briandowns/spinner#available-character-sets
+		// ⣾ ⣷ ⣽ ⣻ ⡿
+		spinnerStyle = spinner.CharSets[11]
+	}
```
**question:** should we bypass `briandowns/spinner` all together if disabled?

Looking under the covers, `briandowns/spinner` requires a spinner character set in order to do anything.  In my mind, I was originally thinking `Working...` message would be treated as a default prefix if `StartProgressIndicator()` was called; not necessarily combined with a provided prefix.

As I mentioned in one of my other comments, `spinner` will still erase and redraw the screen as it doesn't appear to determine if an existing line matches the output.

BagToad replied:
I'm assuming that your concern with redrawing relates to the effect on screen readers, but it would be good if you could clarify that I'm reading between the lines correctly. So, I tested with Mac's screen reader and didn't see any issues with redrawing - the screen reader doesn't repeat the message on every redraw - it only speaks it once.

> question: should we bypass briandowns/spinner all together if disabled?

Could you talk more about your motivation behind this question? I'm not clear on the value proposition of bypassing it.

I'm also not clear on what issues you are alluding to relating to the usage of the library, since my testing showed no apparent impairment from the redrawing.

andyfeller replied:
💯 You are correct, I was unsure about the impact on screen readers from redrawing the screen.  I'm glad to hear how Mac Voiceover behaves though I'm unsure how Windows + NVDA or Ubuntu + Orca will. 🤔

### Technical deep dive

Whenever `briandowns/spinner` is started and stopped from `IOStreams.StartProgressIndicatorWithLabel()` and `IOStreams.StopProgressIndicator()`, it is [erasing the screen](https://github.com/briandowns/spinner/blob/55430861f77b20de3f456775b00c706519464f36/spinner.go#L451-L480).

In our case, we have a single word in the spinner set, so it is just redrawing the same message over and over.  Once the progress indicator stops, all of the indicator messages are erased.

<details>
<summary>
  <b>Expand for <code>Spinner.erase()</code> logic</b>
</summary>
<p>

```go
// erase deletes written characters on the current line.
// Caller must already hold s.lock.
func (s *Spinner) erase() {
	n := utf8.RuneCountInString(s.lastOutputPlain)
	if runtime.GOOS == "windows" && !isWindowsTerminalOnWindows {
		clearString := "\r" + strings.Repeat(" ", n) + "\r"
		fmt.Fprint(s.Writer, clearString)
		s.lastOutputPlain = ""
		return
	}

	numberOfLinesToErase := computeNumberOfLinesNeededToPrintString(s.lastOutputPlain)

	// Taken from https://en.wikipedia.org/wiki/ANSI_escape_code:
	// \r     - Carriage return - Moves the cursor to column zero
	// \033[K - Erases part of the line. If n is 0 (or missing), clear from
	// cursor to the end of the line. If n is 1, clear from cursor to beginning
	// of the line. If n is 2, clear entire line. Cursor position does not
	// change.
	// \033[F - Go to the beginning of previous line
	eraseCodeString := strings.Builder{}
	// current position is at the end of the last printed line. Start by erasing current line
	eraseCodeString.WriteString("\r\033[K") // start by erasing current line
	for i := 1; i < numberOfLinesToErase; i++ {
		// For each additional lines, go up one line and erase it.
		eraseCodeString.WriteString("\033[F\033[K")
	}
	fmt.Fprint(s.Writer, eraseCodeString.String())
	s.lastOutputPlain = ""
}
```

</p>
</details>

BagToad replied:
Resolving because I've made some of these changes - please have another look and let me know what you think ❤️

andyfeller reviewed (commented):
Thanks for getting this pull request opened so quickly, @BagToad! ❤️ 

I see what you mean about the new experience from these efforts, which I agree we likely want something different. 🤔 Honestly, I think we should probably bypass `spinner` all together if disabled.  

At most, we could duplicate the looping effort to print a message every X duration to stderr based on the logic already being used.

Thoughts?

andyfeller commented:
## User Experience feedback

1. I think `Working...` should only show up if a progress indicator is used without a label _(`StartProgressIndicator()`)_; otherwise use the label provided

   So instead of:

   ```shell
   Downloading Working...
   ```

   we should see

   ```shell
   Downloading
   ````

1. I think `spinner` is redrawing the screen, which I'm unsure is something `gh` should avoid...

1. ...but I suppose there is a question whether progress indicators should stay on the screen once the progress indicator is stopped 🤔 

1. I wonder if we should use the [actual ellipsis Unicode character](https://www.compart.com/en/unicode/U+2026) 🤔 _(not the end of the world but just something that popped in my head)_

BagToad committed:
fix(iostreams): spinner env var is now in factory

- Move env var evaluation to the factory defaults.
- Use only the label for the spinner instead of adding
"working..." every time. "working..." remains the default
when no label is provided.

BagToad committed:
feat(iostreams): textual progress indicator does not clear screen

This bypasses the `spinner` package for the textual progress indicator for users
with the spinner disabled out of concerns for accessibility, specifically
with screen readers:

- The `spinner` package will continuously re-draw the screen. I wasn't
able to have this cause problems with my Mac screen reader, but it's
nonetheless a concern that other screen readers may not handle this
screen re-drawing well.
- The `spinner` package clears any progress indicator messages from the screen
when stopping the progress indicator or changing its label.
This is a problem because it interrupts screen readers and leaves no way
to recover what the loading message was by scrolling up in the terminal.

NOTE: this new implementation still interrupts the screen reader when
the a new label is printed, but it does not clear the screen. This makes
the loading messages recoverable, at least.

andyfeller commented on pkg/iostreams/iostreams.go:
```diff
@@ -305,6 +321,24 @@ func (s *IOStreams) StartProgressIndicatorWithLabel(label string) {
 	s.progressIndicator = sp
 }
 
+func (s *IOStreams) startTextualProgressIndicator(label string) {
+	s.progressIndicatorMu.Lock()
+	defer s.progressIndicatorMu.Unlock()
+
+	// Default label when spinner disabled is "Working..."
+	if label == "" {
+		label = "Working..."
+	}
+
+	// Add an ellipsis to the label if it doesn't already have one.
+	ellipsis := "..."
+	if !strings.HasSuffix(label, ellipsis) {
+		label = label + ellipsis
+	}
+
+	fmt.Fprintf(s.ErrOut, "%s%s", s.ColorScheme().Cyan(label), "\n")
+}
```
**thought:** I was going to ask about changes to `StopProgressIndicator()`, but I realized its unnecessary as no `IOStreams.progressIndicator` needs to be stopped.

BagToad replied:
Yeah, I had the same thought. Maybe I should comment this?

BagToad replied:
Commented in 056d292f2635395bdbdc79e9c95fee84bae61724

andyfeller commented on pkg/iostreams/iostreams.go:
```diff
@@ -305,6 +321,24 @@ func (s *IOStreams) StartProgressIndicatorWithLabel(label string) {
+		label = "Working..."
+	}
+
+	// Add an ellipsis to the label if it doesn't already have one.
+	ellipsis := "..."
+	if !strings.HasSuffix(label, ellipsis) {
+		label = label + ellipsis
+	}
```
**praise:** this is a nice touch, the text equivalent of indicating loading

andyfeller commented on pkg/cmd/factory/default_test.go:
```diff
@@ -432,6 +432,56 @@ func Test_ioStreams_prompt(t *testing.T) {
+					t.Setenv(k, v)
+				}
+			}
+			f := New("1")
```
**thought:** Nothing to change here, but this `New("1")` was a bit confusing until I realized this was the `Factory` initializer:

https://github.com/cli/cli/blob/f283d6d11ce1c3d15c3cf55a93d2c1a6fa1fcdd6/pkg/cmd/factory/default.go#L29-L47

In retrospect, I might like seeing this eventually renamed as it looks like a native built-in. $0.02

BagToad replied:
Makes sense, but yeah I'll leave that for another time 🙇

andyfeller commented on pkg/cmd/factory/default_test.go:
```diff
@@ -432,6 +432,56 @@ func Test_ioStreams_prompt(t *testing.T) {
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.env != nil {
+				for k, v := range tt.env {
+					t.Setenv(k, v)
+				}
+			}
```
**question:** does the zero value of `map` mean the `nil` logic isn't needed?

```suggestion
			for k, v := range tt.env {
				t.Setenv(k, v)
			}
```

BagToad replied:
Looks like it! Thanks 😁 

Fixed in dabb29bd361dc29a80b2f666271a28cc0a29f599

andyfeller commented on pkg/iostreams/iostreams_progress_indicator_test.go:
```diff
@@ -0,0 +1,254 @@
+//go:build !windows
```
**question:** This test excludes Windows due to problems with pseudo terminal + `syscall` package, right?

BagToad replied:
Yes, it is the `hinshun/vt10x` that fails on Windows. Same problems with our prompter tests implementation.

andyfeller commented on pkg/iostreams/iostreams_progress_indicator_test.go:
```diff
@@ -0,0 +1,254 @@
+func TestStartProgressIndicatorWithLabel(t *testing.T) {
+	osOut := os.Stdout
+	defer func() { os.Stdout = osOut }()
+	// Why do we need a channel in these tests to implement a timeout instead of
+	// relying on expect's timeout?
+	//
+	// Well, expect's timeout is based on the maximum time of a single read
+	// from the console. This works in cases like prompting where we block
+	// waiting for input because the console is not ready to be read.
+	// But in this case, we are not blocking waiting for input and stdout
+	// can be constantly read. This means the timeout will never be reached
+	// in the event of a expectation failure.
+	// To fix this, we need to implement our own timeout that is based
+	// specifically on the total time spent reading the console and waiting
+	// for the target string instead of the max time for a single read
+	// from the console.
```
**praise:** this is really helpful for future maintainence!

andyfeller commented on pkg/iostreams/iostreams_progress_indicator_test.go:
```diff
@@ -0,0 +1,254 @@
+		}
+	})
+
+	t.Run("multiple indicators with GH_SPINNER_DISABLED shows current label", func(t *testing.T) {
+		console := newTestVirtualTerminal(t)
+		io := newTestIOStreams(t, console, true)
+		progressIndicatorLabel1 := "downloading happiness"
+		progressIndicatorLabel2 := "downloading sadness"
+		done := make(chan error)
+		go func() {
+			_, err := console.ExpectString(progressIndicatorLabel1 + "...")
+			require.NoError(t, err)
+			_, err = console.ExpectString(progressIndicatorLabel2 + "...")
+			done <- err
+		}()
+		io.StartProgressIndicatorWithLabel(progressIndicatorLabel1)
+		defer io.StopProgressIndicator()
+		io.StartProgressIndicatorWithLabel(progressIndicatorLabel2)
+
+		select {
+		case err := <-done:
+			require.NoError(t, err)
+		case <-time.After(2 * time.Second):
+			t.Fatal("Test timed out waiting for progress indicator")
+		}
+	})
```
**question:** can there be multiple, concurrent progress indicators or is this really updating / changing the progress indicator?

I ask because the test name made me think maybe concurrent indicators could be happening but I realize `StartProgressIndicatorWithLabel()` returns early and only updates the prefix:

https://github.com/cli/cli/blob/f283d6d11ce1c3d15c3cf55a93d2c1a6fa1fcdd6/pkg/iostreams/iostreams.go#L289-L309

andyfeller replied:
**praise:** I appreciate the thoroughness of this test! ✨

BagToad replied:
It's really only changing the current progress indicator. 

I think this is a good point. The test name should indicate that it's really just multiple calls to the same progress indicator. I'll fix this.

BagToad replied:
I updated this in 290e78c904dacc0ef7df727b13c138be251781ae. Let me know if you think it still needs more work 🙂

andyfeller commented on pkg/cmd/factory/default.go:
```diff
@@ -283,6 +284,12 @@ func ioStreams(f *cmdutil.Factory) *iostreams.IOStreams {
 		io.SetNeverPrompt(true)
 	}
 
+	ghSpinnerDisabledValue, ghSpinnerDisabledIsSet := os.LookupEnv("GH_SPINNER_DISABLED")
+	falseyValues := []string{"false", "0", "no", ""}
+	if ghSpinnerDisabledIsSet && !slices.Contains(falseyValues, ghSpinnerDisabledValue) {
+		io.SetSpinnerDisabled(true)
+	}
```
**suggest:** I think the only missing from this PR is an update to `gh environment` help topic [here](https://github.com/cli/cli/blob/f283d6d11ce1c3d15c3cf55a93d2c1a6fa1fcdd6/pkg/cmd/root/help_topic.go#L41-L118) for the new environment variable.

BagToad replied:
Done in 1dbb01a1d1d0117f75aeb1ae011bbe53c92969bf - please have a look and let me know if you have feedback :)

andyfeller reviewed (changes requested):
Everything looks really good especially with the PTY test!  The only actual change requested is making sure the new env var shows up in `gh environment`; everything is just random thoughts that don't really block this PR.

BagToad committed:
test(iostreams): update test description

BagToad committed:
doc(iostreams): comment behavior of textual progress indicator

BagToad committed:
doc(help): add documentation for GH_SPINNER_DISABLED

BagToad committed:
test(factory): remove needless nil check

andyfeller commented:
> Instead of `Working...`, the respective operation e.g.:
> 
> * `Searching...`
> * `Finding assets to download...`
> * `Downloading...`
> * etc.
> 
> seem more appropriate here.

I would like to follow up this issue with a more extensive realignment of progress indicator language to avoid the default `Working...` wherever possible, @iamazeem  💯

andyfeller commented on pkg/iostreams/iostreams.go:
```diff
@@ -292,6 +292,10 @@ func (s *IOStreams) StartProgressIndicatorWithLabel(label string) {
 	}
 
 	if s.spinnerDisabled {
+		// If the spinner is disabled, simply print a
+		// textual progress indicator and return.
+		// This means that s.ProgressIndicator will be nil.
+		// See also: the comment on StopProgressIndicator()
```
❤️

andyfeller reviewed (approved):
![shipit-squirrel-fireworks](https://github.com/user-attachments/assets/18896359-b1c5-4547-a5bb-47284e711e2a)
